### PR TITLE
Save parameter value state when you switch to text mode and back in query builder

### LINKED ISSUE #1982: Feature request: Save parameter value state when you switch to text mode and back in query builder
### Similar requests

- [X] I have searched and found no existing similar requests

### How are you using Legend?

Legend Studio

### What problems are you trying to solve?

When you have parameters in a query and you provided a value for them for execution The state of these parameter values is not preserved when you go to text mode and come back to the query builder state.

### Describe the solution you would like to see implemented

We need to preserve the state of paramter values entered by user when they enter text mode and come back to query state

### Describe alternatives you have considered

_No response_

### Documentation, Design, Adoption, Migration Strategy

_No response_

### Contribution

- [X] I would like to work on this feature

### PR DESCRIPTION
**Pull request by gayathrir11**
## Summary

closes #1982

## How did you test this change?

- [x] Test(s) added
- [x] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

https://user-images.githubusercontent.com/87973126/224931628-d08fbb9f-4bfc-4dfe-a789-f210ac725a1a.mp4

### CODE REVIEW

File: packages/legend-query-builder/src/components/__tests__/QueryBuilderVariables.test.tsx [hunk 1]
```diff
@@ -287,3 +287,54 @@ test(integrationTest('Query builder parameter default values'), async () => {
...
+    );
+    expect(getByText(parameterPanel, 'var_1')).not.toBeNull();
+    fireEvent.click(renderResult.getByText('Run Query'));
+    let executeDialog = await waitFor(() => renderResult.getByRole('dialog'));
+    expect(getByText(executeDialog, 'Set Parameter Values'));
+
+    const parameterValue = getByText(executeDialog, 'var_1');
+
+    // Here we mimic the toggling to text mode by calling the rebuildWithQuery function.
+    await act(async () => {
```
Comment (xannem) [inline packages/legend-query-builder/src/components/__tests__/QueryBuilderVariables.test.tsx:323]: Would it be possible to mimic clicking the button by title 'View Query in Pure' and then clicking Close button?

File: packages/legend-query-builder/src/stores/QueryBuilderStateBuilder.ts [hunk 2]
```diff
@@ -608,14 +609,26 @@ export const processParameters = (
...
+        checkIfEquivalent(ps.parameter, parameter),
+      ) === undefined
+    ) {
+      const parameterState = new LambdaParameterState(
+        parameter,
+        queryBuilderState.observableContext,
+        queryBuilderState.graphManagerState.graph,
+      );
+      parameterState.mockParameterValue();
+      queryParameterState.addParameter(parameterState);
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/stores/QueryBuilderStateBuilder.ts:630]: We could be altering the query with this logic as we could move around how we add the parameter states to querybuilder.parametersState. 
i.e `test: String[1]. name:String[1]` -> (text mode edit) `test2: string[1], name:String[1]`
With this logic i believe we would rewrite the query parameter to `name: String[1], test2: String[1]` when we change querystate to lambda. 
Can't we just for do the `forEach` on each parameter and if we find a similar one i.e `checkIfEquivalent` in the old state we use that one and add that to the parameter state, if not we create a new one and add?

File: packages/legend-query-builder/src/components/__tests__/QueryBuilderVariables.test.tsx (hunk 1 above)
Comment (MauricioUyaguari) [reply packages/legend-query-builder/src/components/__tests__/QueryBuilderVariables.test.tsx:323]: lets try this, we might face issues with Monaco editor and the mocking of functions, if so lets add that to the comments on why we don't mock monaco and the network calls. I imagine something will become too expensive :)

**Review by MauricioUyaguari — CHANGES_REQUESTED** [review]

**Review by xannem — APPROVED** [review]

File: packages/legend-query-builder/src/components/__tests__/QueryBuilderVariables.test.tsx [hunk 3]
```diff
@@ -287,3 +291,67 @@ test(integrationTest('Query builder parameter default values'), async () => {
...
+    MockedMonacoEditorInstance.onDidFocusEditorWidget.mockReturnValue(null);
+
+    // Here we mimic the toggling to text mode.
+    fireEvent.click(renderResult.getByTitle('View Query in Pure'));
+    const lambdaEditor = await waitFor(() => renderResult.getByRole('dialog'));
+    const MOCK__globalCompileInFormModeFn = createMock();
+    queryBuilderState.graphManagerState.graphManager.pureCodeToLambda =
+      MOCK__globalCompileInFormModeFn;
+    MOCK__globalCompileInFormModeFn.mockResolvedValue(
+      create_RawLambda(param1Lambda.parameters, param1Lambda.body),
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/components/__tests__/QueryBuilderVariables.test.tsx:342]: should we enhance the test to actually change the lambda parameters to mimic the change we are doing here ?

**Review by MauricioUyaguari — COMMENTED** [review]
will merge but can you look into enhancing the test a bit more. Thank!

**Review by MauricioUyaguari — APPROVED** [review]

File: packages/legend-art/src/testMocks/MonacoEditorMockUtils.tsx [hunk 4]
```diff
@@ -86,6 +87,7 @@ export const MockedMonacoEditorAPI = {
   setModelLanguage: createMock(),
   defineTheme: createMock(),
   EndOfLinePreference: MockedMonacoEditorEndOfLinePreference,
+  removeAllMarkers: createMock(),
```
Comment (akphi) [inline packages/legend-art/src/testMocks/MonacoEditorMockUtils.tsx:90]: why are we changing this, is this problem with rebasing?

File: packages/legend-query-builder/src/stores/QueryBuilderState.ts [hunk 5]
```diff
@@ -301,7 +303,7 @@ export abstract class QueryBuilderState implements CommandRegistrar {
     this.resultState = resultState;
   }
 
-  resetQueryContent(): void {
+  resetQueryBodyContent(): void {
```
[the remark this answers is not in the document]
Comment (akphi) [reply packages/legend-query-builder/src/stores/QueryBuilderState.ts:306]: As discussed, let's revert most of the changes, (also remember to revert the `readonly` changes you made up the top of this file). We want to clone the param state (which you did), but want to be a little more explicit at this level, we can modify `rebuildWithQuery` to take advanced options to preserve the param states. Same goes for `resetQueryResult` in https://github.dev/finos/legend-studio/pull/1984

**Review by akphi — CHANGES_REQUESTED** [review]

File: packages/legend-art/src/testMocks/MonacoEditorMockUtils.tsx (hunk 4 above)
Comment (gayathrir11) [reply packages/legend-art/src/testMocks/MonacoEditorMockUtils.tsx:90]: I was trying to monaco in the test so i needed these functions

**Pull request merged**